New PR for Secrets of query performance blog post (#3185)
PostHog/posthog.com: contents/blog/secrets-of-posthog-query-performance.md

---

> date: "2022-03-22" # :TODO

Author: ```suggestion
date: "2022-03-22"
```

title: "Secrets of PostHog query performance"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true

> featuredImage: ../images/blog/launch-week-teaser.jpeg # :TODO:

Reviewer: cc @lottiecoxon this needs changing. You mentioned under issue you have an image planned.

Author: It's ok, it's on Figma already I believe. I need to export, so it's an action for me.

---

> date: "2022-03-22" # :TODO

Author: ```suggestion
date: "2022-03-22"
```

> title: "Secrets of PostHog query performance"

Author: ```suggestion
title: "The secrets of PostHog query performance"
```

> title: "Secrets of PostHog query performance"

Author: ```suggestion
title: "The secrets of PostHog query performance"
```

featuredImageType: full
categories: ["Engineering", "Product updates", "Launch week"]
author: ["karl-aksel-puulmann"]
---

We want PostHog to become the first choice for product analytics at any scale. To do that, users should have a smooth experience exploring their product data - including not waiting minutes for queries to load.

In this post, I’m going to break down the why’s and how’s of several major performance improvements we've made in the last few months to achieve that goal.

> This post is part of our [A Universe of New Features launch week](/blog/launch-week-universe-of-new-features) series

## How does querying work within PostHog?

The free-form querying experience in PostHog allows you to ask questions about your Trends, Funnels, Retention, and Cohorts with complicated filtering to top it off.

These queries are processed by [ClickHouse](https://clickhouse.com/), where event, user, and group data is stored in a raw format without any preaggregation. ClickHouse excels at analyzing billions of events in seconds even with complex filtering.

However, as awesome as ClickHouse is, nothing is without sharp edges and trade-offs.

## Speeding up property filtering by 25x

> PostHog allows users to send and analyze arbitrary number of event and user properties with their data. We store this data as JSON-encoded strings in our tables.

Reviewer: ```suggestion
PostHog allows users to send and analyze an arbitrary number of event and user properties with their data. We store these properties as JSON-encoded string columns in our tables.
```

> PostHog allows users to send and analyze arbitrary number of event and user properties with their data. We store this data as JSON-encoded strings in our tables.

Author: This phrase is a bit confusing to me. Should read "an arbitrary number of events..."

```sql
ORDER BY (
    project_id,
    toDate(timestamp),
    event,
    cityHash64(distinct_id),
    cityHash64(uuid)
)
```

This resulted in a roughly 23% query speed up on average. The best trick for performance optimizations is to skip doing unnecessary work.

### Migrating data

Changing how data is sorted on disk is not cheap when you have billions on billions of events. For example, on PostHog Cloud, this took five separate attempts and multiple weeks in total to finish. In addition, PostHog has a lot of self-hosted users at various degrees of scale and technical skill who would need to repeat this process.

To tackle this problem, we ended up building a new [async migrations](/blog/async-migrations) system, which safely runs these long-running operations at scale with the press of a button, while handling common edge cases and keeping the platform up and usable.

## Self-hosted performance

As mentioned, PostHog can be self-hosted by our users. However, getting it working smoothly across a wide range of deployments at scale [keeps our infrastructure team hard at work](/blog/improving-posthog-deployments).

> Some features released in PostHog 1.34.0 which affect performance for self-hosted users are:

Author: ```suggestion
Some features coming in PostHog 1.34.0 (releasing on Thursday) which affect performance for self-hosted users are:
```
